File: sdks/python/src/verdocs/resources/template_documents.py
"""Template document operations (js-sdk: Templates/TemplateDocuments.ts).

Two js-sdk functions here (getTemplateDocumentFile and
getTemplateDocumentThumbnail) target routes that do not exist on the deployed
API; they are ported for shape parity with doc comments stating the reality.
getTemplateDocumentPreviewLink fetches the envelope-documents path for a
template document; the code is the wire truth, so we mirror the anomaly.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Literal

from ..models import Template, TemplateDocument
from .templates import TemplateFile, _file_part, _multipart_request, _multipart_request_async

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_DOCUMENTS_PATH = "/v2/template-documents"


class TemplateDocuments:
    """Template document calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def create(self, template_id: str, file: TemplateFile) -> TemplateDocument:
        """Attach a document to a template. Mirrors js-sdk createTemplateDocument.

        Sends POST /v2/template-documents as multipart with exactly one file
        part named "file" plus a template_id text part (the js-sdk doc tag
        claims a different path; this is the one the code uses and the server
        serves). The document name comes from the uploaded filename, with no
        override, and the API accepts PDF and DOCX only, judged by the part's
        declared content type.

        Example:
            document = template_documents.create(template.id, "/contracts/nda.pdf")

        Args:
            template_id: ID of the template to attach the document to.
            file: The document: a path, raw PDF bytes, an open binary file,
                or an httpx-style (filename, content[, content_type]) tuple.

        Returns:
            The created template document row (not the containing template).

        Raises:
            VerdocsAPIError: The API rejected the upload.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.createTemplateDocument
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        parts = [("file", _file_part(file))]
        response = _multipart_request(
            self._endpoint, "POST", _DOCUMENTS_PATH, data={"template_id": template_id}, files=parts
        )
        return TemplateDocument.model_validate(response.json())

    def delete(self, document_id: str) -> Template:
        """Delete a template document via DELETE /v2/template-documents/{document_id}.

        Mirrors js-sdk deleteTemplateDocument. The server answers with the
        remaining deep template (documents, fields, roles), not the status
        string the js-sdk doc tag claims, so this returns the updated template.

        Args:
            document_id: ID of the document to delete.

        Returns:
            The template as it stands after the deletion.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.deleteTemplateDocument
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request("DELETE", f"{_DOCUMENTS_PATH}/{document_id}")
        return Template.model_validate(response.json())

    def get(self, document_id: str) -> TemplateDocument:
        """Get a template document's metadata via GET /v2/template-documents/{document_id}.

        Mirrors js-sdk getTemplateDocument. Non-creators (e.g. org
        collaborators) get only the metadata they are allowed to view. The
        server stringifies the JSON itself and sends it with a text/html
        content type; we parse the body as JSON regardless.

        Args:
            document_id: ID of the document to fetch.

        Returns:
            The document metadata.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.getTemplateDocument
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/{document_id}")
        return TemplateDocument.model_validate(response.json())

    def download(self, document_id: str) -> bytes:
        """Download a template document's raw bytes. Mirrors js-sdk downloadTemplateDocument.

        Calls GET /v2/template-documents/{document_id}?type=file. The response
        carries the document's stored mime type; there is no redirect.

        Args:
            document_id: ID of the document to download.

        Returns:
            The document content.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.downloadTemplateDocument
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/{document_id}", params={"type": "file"})
        return response.content

    def get_download_link(self, document_id: str) -> str:
        """Get a pre-signed download link for a template document.

        Mirrors js-sdk getTemplateDocumentDownloadLink (minus its unused
        envelope_id parameter). Calls
        GET /v2/template-documents/{document_id}?type=download; the server
        answers with a signed Cloudfront URL as a bare string (text/html
        content type, not JSON), attachment disposition, one hour expiry.
        Fetch the link immediately and never share it.

        Args:
            document_id: ID of the document to link to.

        Returns:
            The signed download URL.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.getTemplateDocumentDownloadLink
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/{document_id}", params={"type": "download"})
        return response.text

    def get_preview_link(self, document_id: str) -> str:
        """Get a pre-signed preview link (inline disposition) for a template document.

        Mirrors js-sdk getTemplateDocumentPreviewLink (minus its unused
        envelope_id parameter), which fetches
        GET /v2/envelope-documents/{document_id}?type=preview: the ENVELOPE
        document family, not the template one. The code is the wire truth, so
        we mirror it; the template-family route
        GET /v2/template-documents/{id}?type=preview also serves preview
        links. The URL comes back as a bare string and expires quickly, so use
        it immediately and never share it.

        Args:
            document_id: ID of the document to link to.

        Returns:
            The signed preview URL.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.getTemplateDocumentPreviewLink
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request("GET", f"/v2/envelope-documents/{document_id}", params={"type": "preview"})
        return response.text

    def get_file(self, template_id: str, document_id: str) -> bytes:
        """Get (binary download) a file attached to a template. Mirrors js-sdk getTemplateDocumentFile.

        Dead on the deployed API: the
        GET /v2/templates/{template_id}/documents/{document_id} route it
        targets exists nowhere in the API, so every call fails with a 404.
        download() is the working equivalent. Ported for shape parity only.

        Args:
            template_id: ID of the template the document belongs to.
            document_id: ID of the document to download.

        Returns:
            The document content, if the API ever ships the route.

        Raises:
            NotFoundError: Always today; the deployed API has no such route.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.getTemplateDocumentFile
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request(
            "GET", f"/v2/templates/{template_id}/documents/{document_id}", params={"file": "true"}
        )
        return response.content

    def get_thumbnail(self, template_id: str, document_id: str) -> bytes:
        """Get a thumbnail image for a template document. Mirrors js-sdk getTemplateDocumentThumbnail.

        Dead on the deployed API: the
        GET /v2/templates/{template_id}/documents/{document_id} route it
        targets exists nowhere in the API, so every call fails with a 404.
        get_page_display_uri(document_id, "thumb") is the working equivalent.
        Ported for shape parity only.

        Args:
            template_id: ID of the template the document belongs to.
            document_id: ID of the document to thumbnail.

        Returns:
            The thumbnail content, if the API ever ships the route.

        Raises:
            NotFoundError: Always today; the deployed API has no such route.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.getTemplateDocumentThumbnail
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request(
            "GET", f"/v2/templates/{template_id}/documents/{document_id}", params={"thumbnail": "true"}
        )
        return response.content

    def get_page_display_uri(
        self, document_id: str, page: int | Literal["thumb"], variant: Literal["original", "tagged"] = "original"
    ) -> str:
        """Get a display URL for one page of a template document, rendered server-side as a PNG.

        Mirrors js-sdk getTemplateDocumentPageDisplayUri. Calls
        GET /v2/template-documents/page-image/{document_id}/{variant}/{page}
        and returns a signed Cloudfront URL as a bare string. Pages are
        0-based; pass "thumb" for the thumbnail rendition (the js-sdk types
        page as a number, but the server accepts the literal). These renders
        are for display only: they are not legally binding documents and carry
        no participant metadata; download() returns the original asset.

        Example:
            url = template_documents.get_page_display_uri(document.id, 0)
            thumb = template_documents.get_page_display_uri(document.id, "thumb")

        Args:
            document_id: ID of the document to render.
            page: 0-based page number, or "thumb" for the thumbnail.
            variant: "original" (default) or "tagged".

        Returns:
            The signed page image URL.

        Raises:
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation templateDocument.getTemplateDocumentPageDisplayUri
        @sdkGroup TemplateDocument
        @sdkPage Endpoints
        """
        response = self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/page-image/{document_id}/{variant}/{page}")
        return response.text


class AsyncTemplateDocuments:
    """Template document calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def create(self, template_id: str, file: TemplateFile) -> TemplateDocument:
        """Attach a document to a template. Mirrors js-sdk createTemplateDocument.

        Sends POST /v2/template-documents as multipart with exactly one file
        part named "file" plus a template_id text part (the js-sdk doc tag
        claims a different path; this is the one the code uses and the server
        serves). The document name comes from the uploaded filename, with no
        override, and the API accepts PDF and DOCX only, judged by the part's
        declared content type.

        Example:
            document = await template_documents.create(template.id, "/contracts/nda.pdf")

        Args:
            template_id: ID of the template to attach the document to.
            file: The document: a path, raw PDF bytes, an open binary file,
                or an httpx-style (filename, content[, content_type]) tuple.

        Returns:
            The created template document row (not the containing template).

        Raises:
            VerdocsAPIError: The API rejected the upload.
            VerdocsConnectionError: The request never reached the API.
        """
        parts = [("file", _file_part(file))]
        response = await _multipart_request_async(
            self._endpoint, "POST", _DOCUMENTS_PATH, data={"template_id": template_id}, files=parts
        )
        return TemplateDocument.model_validate(response.json())

    async def delete(self, document_id: str) -> Template:
        """Delete a template document via DELETE /v2/template-documents/{document_id}.

        Mirrors js-sdk deleteTemplateDocument. The server answers with the
        remaining deep template (documents, fields, roles), not the status
        string the js-sdk doc tag claims, so this returns the updated template.

        Args:
            document_id: ID of the document to delete.

        Returns:
            The template as it stands after the deletion.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("DELETE", f"{_DOCUMENTS_PATH}/{document_id}")
        return Template.model_validate(response.json())

    async def get(self, document_id: str) -> TemplateDocument:
        """Get a template document's metadata via GET /v2/template-documents/{document_id}.

        Mirrors js-sdk getTemplateDocument. Non-creators (e.g. org
        collaborators) get only the metadata they are allowed to view. The
        server stringifies the JSON itself and sends it with a text/html
        content type; we parse the body as JSON regardless.

        Args:
            document_id: ID of the document to fetch.

        Returns:
            The document metadata.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/{document_id}")
        return TemplateDocument.model_validate(response.json())

    async def download(self, document_id: str) -> bytes:
        """Download a template document's raw bytes. Mirrors js-sdk downloadTemplateDocument.

        Calls GET /v2/template-documents/{document_id}?type=file. The response
        carries the document's stored mime type; there is no redirect.

        Args:
            document_id: ID of the document to download.

        Returns:
            The document content.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/{document_id}", params={"type": "file"})
        return response.content

    async def get_download_link(self, document_id: str) -> str:
        """Get a pre-signed download link for a template document.

        Mirrors js-sdk getTemplateDocumentDownloadLink (minus its unused
        envelope_id parameter). Calls
        GET /v2/template-documents/{document_id}?type=download; the server
        answers with a signed Cloudfront URL as a bare string (text/html
        content type, not JSON), attachment disposition, one hour expiry.
        Fetch the link immediately and never share it.

        Args:
            document_id: ID of the document to link to.

        Returns:
            The signed download URL.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/{document_id}", params={"type": "download"})
        return response.text

    async def get_preview_link(self, document_id: str) -> str:
        """Get a pre-signed preview link (inline disposition) for a template document.

        Mirrors js-sdk getTemplateDocumentPreviewLink (minus its unused
        envelope_id parameter), which fetches
        GET /v2/envelope-documents/{document_id}?type=preview: the ENVELOPE
        document family, not the template one. The code is the wire truth, so
        we mirror it; the template-family route
        GET /v2/template-documents/{id}?type=preview also serves preview
        links. The URL comes back as a bare string and expires quickly, so use
        it immediately and never share it.

        Args:
            document_id: ID of the document to link to.

        Returns:
            The signed preview URL.

        Raises:
            NotFoundError: No visible document has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request(
            "GET", f"/v2/envelope-documents/{document_id}", params={"type": "preview"}
        )
        return response.text

    async def get_file(self, template_id: str, document_id: str) -> bytes:
        """Get (binary download) a file attached to a template. Mirrors js-sdk getTemplateDocumentFile.

        Dead on the deployed API: the
        GET /v2/templates/{template_id}/documents/{document_id} route it
        targets exists nowhere in the API, so every call fails with a 404.
        download() is the working equivalent. Ported for shape parity only.

        Args:
            template_id: ID of the template the document belongs to.
            document_id: ID of the document to download.

        Returns:
            The document content, if the API ever ships the route.

        Raises:
            NotFoundError: Always today; the deployed API has no such route.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request(
            "GET", f"/v2/templates/{template_id}/documents/{document_id}", params={"file": "true"}
        )
        return response.content

    async def get_thumbnail(self, template_id: str, document_id: str) -> bytes:
        """Get a thumbnail image for a template document. Mirrors js-sdk getTemplateDocumentThumbnail.

        Dead on the deployed API: the
        GET /v2/templates/{template_id}/documents/{document_id} route it
        targets exists nowhere in the API, so every call fails with a 404.
        get_page_display_uri(document_id, "thumb") is the working equivalent.
        Ported for shape parity only.

        Args:
            template_id: ID of the template the document belongs to.
            document_id: ID of the document to thumbnail.

        Returns:
            The thumbnail content, if the API ever ships the route.

        Raises:
            NotFoundError: Always today; the deployed API has no such route.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request(
            "GET", f"/v2/templates/{template_id}/documents/{document_id}", params={"thumbnail": "true"}
        )
        return response.content

    async def get_page_display_uri(
        self, document_id: str, page: int | Literal["thumb"], variant: Literal["original", "tagged"] = "original"
    ) -> str:
        """Get a display URL for one page of a template document, rendered server-side as a PNG.

        Mirrors js-sdk getTemplateDocumentPageDisplayUri. Calls
        GET /v2/template-documents/page-image/{document_id}/{variant}/{page}
        and returns a signed Cloudfront URL as a bare string. Pages are
        0-based; pass "thumb" for the thumbnail rendition (the js-sdk types
        page as a number, but the server accepts the literal). These renders
        are for display only: they are not legally binding documents and carry
        no participant metadata; download() returns the original asset.

        Example:
            url = await template_documents.get_page_display_uri(document.id, 0)
            thumb = await template_documents.get_page_display_uri(document.id, "thumb")

        Args:
            document_id: ID of the document to render.
            page: 0-based page number, or "thumb" for the thumbnail.
            variant: "original" (default) or "tagged".

        Returns:
            The signed page image URL.

        Raises:
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", f"{_DOCUMENTS_PATH}/page-image/{document_id}/{variant}/{page}")
        return response.text

File: sdks/csharp/src/Verdocs.Sdk/Resources/TemplateDocuments.cs
using System.Globalization;
using System.Net.Http.Headers;
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// Template document (attachment) calls, reached through the endpoint's TemplateDocuments
/// property. Documents are listed under their template; call
/// <see cref="Templates.GetAsync"/> to enumerate them.
/// </summary>
public sealed class TemplateDocuments
{
    private readonly VerdocsEndpoint _endpoint;

    internal TemplateDocuments(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Attaches a document to a template. The upload is multipart: one file part named "file"
    /// plus the template ID as a text part. The document takes its name from the uploaded
    /// file name; there is no name override. PDF and DOCX are accepted (DOCX is converted to
    /// PDF server-side, with the original retained), up to 25 MB.
    ///
    /// <example>
    /// <code>
    /// await using var stream = File.OpenRead("bill-of-sale.pdf");
    /// var document = await endpoint.TemplateDocuments.CreateAsync(
    ///     "d2338742-f3a1-465b-8592-806587413cc1",
    ///     new TemplateFileUpload { Content = stream, FileName = "bill-of-sale.pdf", ContentType = "application/pdf" });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template to attach the document to.</param>
    /// <param name="file">The file to upload.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new template document.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the file was an unsupported type.</exception>
    /// <sdkOperation>templateDocument.createTemplateDocument</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<TemplateDocument> CreateAsync(string templateId, TemplateFileUpload file, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        ArgumentNullException.ThrowIfNull(file);

        // Names are pre-quoted because .NET only quotes multipart part names when it must,
        // while browsers (and so the API's usual traffic) always quote them. The part content
        // type matters: the server runs its PDF/DOCX check against the declaration.
        var content = new MultipartFormDataContent();
        var part = new StreamContent(file.Content);
        part.Headers.ContentType = new MediaTypeHeaderValue(file.ContentType);
        content.Add(part, "\"file\"", "\"" + file.FileName + "\"");
        content.Add(new StringContent(templateId), "\"template_id\"");

        return _endpoint.SendAsync<TemplateDocument>(HttpMethod.Post, "/v2/template-documents", content, cancellationToken);
    }

    /// <summary>
    /// Gets a template document's metadata. Non-creators (organization collaborators, for
    /// example) receive only the metadata they are allowed to view. The server sends this
    /// JSON with a text/html content type; the SDK parses the body as JSON regardless.
    ///
    /// <example>
    /// <code>
    /// var document = await endpoint.TemplateDocuments.GetAsync(documentId);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="documentId">The document's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The document's metadata.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the document was not found.</exception>
    /// <sdkOperation>templateDocument.getTemplateDocument</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<TemplateDocument> GetAsync(string documentId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        return _endpoint.SendAsync<TemplateDocument>(
            HttpMethod.Get, "/v2/template-documents/" + Uri.EscapeDataString(documentId), null, cancellationToken);
    }

    /// <summary>
    /// Downloads a template document's file directly. Returns the raw bytes with no redirect;
    /// use this rather than a plain link so the authorization headers are set.
    ///
    /// <example>
    /// <code>
    /// var bytes = await endpoint.TemplateDocuments.DownloadAsync(documentId);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="documentId">The document's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The file's raw bytes.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the document was not found.</exception>
    /// <sdkOperation>templateDocument.downloadTemplateDocument</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<byte[]> DownloadAsync(string documentId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        return _endpoint.SendBytesAsync(
            HttpMethod.Get, "/v2/template-documents/" + Uri.EscapeDataString(documentId) + "?type=file", cancellationToken);
    }

    /// <summary>
    /// Gets a pre-signed download link for a template document (attachment disposition, about
    /// a one hour expiry). The link needs no auth header, so use it immediately and never
    /// share it.
    ///
    /// <example>
    /// <code>
    /// var link = await endpoint.TemplateDocuments.GetDownloadLinkAsync(documentId);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="documentId">The document's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The signed URL, as a bare string.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the document was not found.</exception>
    /// <sdkOperation>templateDocument.getTemplateDocumentDownloadLink</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<string> GetDownloadLinkAsync(string documentId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        return _endpoint.SendStringAsync(
            HttpMethod.Get, "/v2/template-documents/" + Uri.EscapeDataString(documentId) + "?type=download", cancellationToken);
    }

    /// <summary>
    /// Gets a pre-signed preview link (inline disposition). Anomaly preserved from the
    /// js-sdk: this fetches the envelope-documents route, not template-documents, so it only
    /// resolves for envelope document IDs. The template-documents equivalent is
    /// GET /v2/template-documents/:id?type=preview, which this SDK does not wrap yet.
    ///
    /// <example>
    /// <code>
    /// var link = await endpoint.TemplateDocuments.GetPreviewLinkAsync(documentId);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="documentId">The document ID to link to.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The signed URL, as a bare string.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the document was not found.</exception>
    /// <sdkOperation>templateDocument.getTemplateDocumentPreviewLink</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<string> GetPreviewLinkAsync(string documentId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        return _endpoint.SendStringAsync(
            HttpMethod.Get, "/v2/envelope-documents/" + Uri.EscapeDataString(documentId) + "?type=preview", cancellationToken);
    }

    /// <summary>
    /// Gets a template document file through the legacy templates route. Dead on the deployed
    /// API: the route does not exist, so every call fails. Ported to match the js-sdk
    /// surface; <see cref="DownloadAsync"/> is the working equivalent. Retirement is pending.
    ///
    /// <example>
    /// <code>
    /// var bytes = await endpoint.TemplateDocuments.GetFileAsync(templateId, documentId);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template the document belongs to.</param>
    /// <param name="documentId">The document's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The file's raw bytes, if the API ever serves this route.</returns>
    /// <exception cref="VerdocsApiException">Always thrown today; the deployed API has no handler for this call.</exception>
    /// <sdkOperation>templateDocument.getTemplateDocumentFile</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<byte[]> GetFileAsync(string templateId, string documentId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        return _endpoint.SendBytesAsync(
            HttpMethod.Get,
            "/v2/templates/" + Uri.EscapeDataString(templateId) + "/documents/" + Uri.EscapeDataString(documentId) + "?file=true",
            cancellationToken);
    }

    /// <summary>
    /// Gets a template document thumbnail through the legacy templates route. Dead on the
    /// deployed API: the route does not exist, so every call fails. Ported to match the
    /// js-sdk surface; the working way to get a thumbnail is the page-image route with the
    /// literal page value "thumb", which this SDK does not wrap yet
    /// (<see cref="GetPageDisplayUriAsync"/> takes numeric pages, matching the js-sdk).
    /// Retirement is pending.
    ///
    /// <example>
    /// <code>
    /// var bytes = await endpoint.TemplateDocuments.GetThumbnailAsync(templateId, documentId);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="templateId">The template the document belongs to.</param>
    /// <param name="documentId">The document's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The thumbnail's raw bytes, if the API ever serves this route.</returns>
    /// <exception cref="VerdocsApiException">Always thrown today; the deployed API has no handler for this call.</exception>
    /// <sdkOperation>templateDocument.getTemplateDocumentThumbnail</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<byte[]> GetThumbnailAsync(string templateId, string documentId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(templateId);
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        return _endpoint.SendBytesAsync(
            HttpMethod.Get,
            "/v2/templates/" + Uri.EscapeDataString(templateId) + "/documents/" + Uri.EscapeDataString(documentId) + "?thumbnail=true",
            cancellationToken);
    }

    /// <summary>
    /// Gets a display URI for one page of a template document. Pages are rendered server-side
    /// to PNG for display only; they are not legally binding and carry no participant
    /// metadata. Get the original file via <see cref="DownloadAsync"/>.
    ///
    /// <example>
    /// <code>
    /// var uri = await endpoint.TemplateDocuments.GetPageDisplayUriAsync(documentId, 0);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="documentId">The document to render.</param>
    /// <param name="page">0-based page number (0-1000).</param>
    /// <param name="variant">"original" or "tagged".</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The signed URL, as a bare string.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the document was not found.</exception>
    /// <sdkOperation>templateDocument.getTemplateDocumentPageDisplayUri</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<string> GetPageDisplayUriAsync(string documentId, int page, string variant = "original", CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        ArgumentOutOfRangeException.ThrowIfNegative(page);
        ArgumentException.ThrowIfNullOrEmpty(variant);
        return _endpoint.SendStringAsync(
            HttpMethod.Get,
            "/v2/template-documents/page-image/" + Uri.EscapeDataString(documentId) + "/" + Uri.EscapeDataString(variant)
                + "/" + page.ToString(CultureInfo.InvariantCulture),
            cancellationToken);
    }

    /// <summary>
    /// Deletes a template document. Returns the remaining deep template: the js-sdk doc tag
    /// claims a bare status string, but its code and the server both return the template.
    ///
    /// <example>
    /// <code>
    /// var template = await endpoint.TemplateDocuments.DeleteAsync(documentId);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="documentId">The document's unique ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The template the document was removed from, with its remaining documents, roles, and fields.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the caller may not edit the template.</exception>
    /// <sdkOperation>templateDocument.deleteTemplateDocument</sdkOperation>
    /// <sdkGroup>TemplateDocument</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Template> DeleteAsync(string documentId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(documentId);
        return _endpoint.SendAsync<Template>(
            HttpMethod.Delete, "/v2/template-documents/" + Uri.EscapeDataString(documentId), null, cancellationToken);
    }
}
